src/main/kotlin/com/exactpro/th2/common/microservice/Main.kt

```diff
@@ -0,0 +1,151 @@
+
+    init {
+        configureShutdownHook(resources, lock, condition, liveness, readiness)
+
+        K_LOGGER.info { "Starting a component" }
+        liveness.enable()
+
+        val applicationFactory = loadApplicationFactory()
+            .also(resources::add)
+        commonFactory = CommonFactory.createFromArguments(*args)
+            .also(resources::add)
+        application = applicationFactory.createApplication(ApplicationContext(commonFactory, resources::add, ::panic))
+            .also(resources::add)
+
+        K_LOGGER.info { "Prepared the '${commonFactory.boxConfiguration.boxName}' component" }
+    }
+
+    fun run() {
+        K_LOGGER.info { "Starting the '${commonFactory.boxConfiguration.boxName}' component" }
+
+        application.start()
+
+        K_LOGGER.info { "Started the '${commonFactory.boxConfiguration.boxName}' component" }
+        readiness.enable()
+
+        awaitShutdown(lock, condition)
+    }
+
+    companion object {
+        private const val APPLICATION_FACTORY_SYSTEM_PROPERTY = "th2.microservice.application-factory"
+        private const val MONITOR_NAME = "microservice_main"
+        private fun loadApplicationFactory(): IApplicationFactory {
+            val instances = ServiceLoader.load(IApplicationFactory::class.java).toList()
+            return when (instances.size) {
+                0 -> error("No instances of ${IApplicationFactory::class.simpleName}")
+                1 -> instances.single().also { single ->
+                    System.getProperty(APPLICATION_FACTORY_SYSTEM_PROPERTY)?.let { value ->
+                        check(value == single::class.qualifiedName) {
+                            "Found instance of ${IApplicationFactory::class.simpleName} mismatches the class specified by $APPLICATION_FACTORY_SYSTEM_PROPERTY system property," +
+                                    "configured: $value, found: ${single::class.qualifiedName}"
+                        }
+                    }
+                }
+
+                else -> {
+                    System.getProperty(APPLICATION_FACTORY_SYSTEM_PROPERTY)?.let { value ->
+                        instances.find { value == it::class.qualifiedName }
+                            ?: error(
+                                "Found instances of ${IApplicationFactory::class.simpleName} mismatches the class specified by $APPLICATION_FACTORY_SYSTEM_PROPERTY system property," +
+                                        "configured: $value, found: ${instances.map { Object::class.qualifiedName }}"
+                            )
+                    } ?: error(
+                        "More than 1 instance of ${IApplicationFactory::class.simpleName} has been found " +
+                                "and $APPLICATION_FACTORY_SYSTEM_PROPERTY system property isn't specified," +
+                                "instances: $instances"
+                    )
+                }
+            }
+        }
+    }
+}
+
+fun main(args: Array<String>) {
```

Comment:
From my own experience, having a main method in another project is not really convenient for understanding what is going on when you run the application...
If we take a look at Spring Boot (https://spring.io/quickstart - step 3) the main method is defined in the application code (not in the library). The library only provides users with a starter method:
```java
SpringApplication.run(DemoApplication.class, args);
```

Probably, we should go with the same approach, in my opinion.
We can still use ServiceLoader mechanism under the hood but move the main method from library to the application code